File: rust/day_9/part_1.rs
use std::fs;

fn solve(input: &str) -> i32 {
    let lines: Vec<&str> = input
        .trim()
        .split('\n')
        .map(|line| line.trim())
        .filter(|line| !line.is_empty())
        .collect();

    let mut tiles = Vec::new();
    for line in lines {
        let parts: Vec<i32> = line.split(',').map(|s| s.parse().unwrap()).collect();
        tiles.push((parts[0], parts[1]));
    }

    let n = tiles.len();
    let mut max_area = 0;

    for i in 0..n {
        for j in (i + 1)..n {
            let (x1, y1) = tiles[i];
            let (x2, y2) = tiles[j];

            let width = (x2 - x1).abs() + 1;
            let height = (y2 - y1).abs() + 1;
            let area = width * height;

            max_area = max_area.max(area);
        }
    }

    max_area
}

fn main() {
    let input = fs::read_to_string("../inputs/day_9.txt").unwrap();
    let result = solve(&input);
    println!("{}", result);
}

File: java/day_11/Part1.java
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class Part1 {
    static Map<String, List<String>> graph;
    static Map<String, Integer> memo;

    public static int count_paths(String node) {
        if (node.equals("out")) {
            return 1;
        }

        if (memo.containsKey(node)) {
            return memo.get(node);
        }

        if (!graph.containsKey(node)) {
            return 0;
        }

        int total = 0;
        for (String neighbor : graph.get(node)) {
            total += count_paths(neighbor);
        }

        memo.put(node, total);
        return total;
    }

    public static int solve(String input) {
        List<String> lines = new ArrayList<>();
        for (String line : input.trim().split("\n")) {
            line = line.trim();
            if (!line.isEmpty()) {
                lines.add(line);
            }
        }

        graph = new HashMap<>();
        for (String line : lines) {
            String[] parts = line.split(":");
            String device = parts[0].trim();
            String[] outputs = parts[1].trim().split("\\s+");
            List<String> outputList = new ArrayList<>();
            for (String output : outputs) {
                outputList.add(output);
            }
            graph.put(device, outputList);
        }

        memo = new HashMap<>();

        return count_paths("you");
    }

    public static void main(String[] args) throws IOException {
        String input = new String(Files.readAllBytes(Paths.get("../inputs/day_11.txt")));
        int result = solve(input);
        System.out.println(result);
    }
}
